src/python/pants/backend/python/macros/python_artifact.py

```diff
@@ -77,7 +78,12 @@ def __init__(self, **kwargs):
             # coerce entry points from Dict[str, List[str]] to Dict[str, Dict[str, str]]
             kwargs["entry_points"] = _normalize_entry_points(kwargs["entry_points"])
 
-        self._kw: Dict[str, Any] = kwargs
+        self._kw: Dict[str, Any] = copy.deepcopy(kwargs)
+        # The kwargs come from a BUILD file, and can contain somewhat arbitrary nested structures,
+        # so we don't have a principled way to make them into a hashable data structure.
+        # Instead we stringify and precompute a hash to use in our own __hash__, since we know
+        # that our
+        self._hash: int = hash(str(kwargs))
```

Comment:
Isn't it the case that we know that the arbitrariness does not extend beyond json serializable bounds? Perhaps this is already just as good, but a sorted-key json serialization hashed might be better.

Reply:
Yeah, I think that's right.

Reply:
In that case we could also recursively replace dicts with FrozenDicts and lists with tuples and be done with it.

Reply:
Erk, this is no good because some code deep inside distutils actually checks that `ext_modules` is a list, and won't accept a tuple. And for all I know there may be other such checks, in either direction. So JSON serialization it is.